File: AppSets-iOS/AppSets/com/app/appsets/im/RabbitMQBroker.swift
//
//  RabbitMQBroker.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/2.
//

import Foundation
import RMQClient


class SimpleRMQConnectionRecovery: NSObject, RMQConnectionRecovery{
    
    var interval: NSNumber!
    
    func recover(_ connection: (any RMQStarter)!, channelAllocator allocator: (any RMQChannelAllocator)!, error: (any Error)!) {
        
    }
    
}

class PrivateRMQConnectionDelegateLogger: RMQConnectionDelegateLogger{
    override func connection(_ connection: RMQConnection!, failedToConnectWithError error: (any Error)!) {
        BrokerTest.Instance.changeOnline(false)
    }
    override func connection(_ connection: RMQConnection!, disconnectedWithError error: (any Error)!) {
        BrokerTest.Instance.changeOnline(false)
    }
    override func recoveredConnection(_ connection: RMQConnection!) {
        BrokerTest.Instance.changeOnline(true)
    }
}


class RabbitMQBroker: MessageBroker {
    
    public static let TAG = "RabbitMQBroker"
    
    public static let MESSAGE_KEY_ON_IM_MESSAGE = "ON_IM_MESSAGE"
    
    typealias C = RabbitMQBrokerConfig
    
    private var connection: RMQConnection?
    
    private var channel: RMQChannel? = nil
    
    private var property: RabbitMqBrokerProperty? = nil
    
    private var queueName: String? = nil
    
    private var routingKey: String? = nil
    
    private var uid: String? = nil
    
    func bootstrap(_ config: RabbitMQBrokerConfig) {
        PurpleLogger.current.d(RabbitMQBroker.TAG, "bootstrap")
        updateUserChannel(config.property)
    }
    
    private func updateUserChannel(_ property: RabbitMqBrokerProperty) {
        PurpleLogger.current.d(RabbitMQBroker.TAG, "updateUserChannel")
        self.property = property
        let userGroupsChanged = true
        let loggedUserUid = LocalAccountManager.Instance.userInfo.uid
        let userChanged = self.uid != loggedUserUid
        self.uid = loggedUserUid
        basicConnect()
        doUserChannelThings(userChanged, userGroupsChanged)
    }
    
    private func basicConnect() {
        if checkConnection() {
            PurpleLogger.current.d(RabbitMQBroker.TAG,"basicConnect! connection is connected, return")
            return
        }
        guard let property = self.property else {
            return
        }
        do {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "startConnection")
            let uri = "amqp://\(property.username):\(property.password )@\(property.host):\(property.port)/\(property.virtualHost)"
            connection = RMQConnection(uri: uri, delegate: PrivateRMQConnectionDelegateLogger())
            connection?.start()
            guard let channel = try connection?.createChannel() else {
                return
            }
            self.channel = channel
            BrokerTest.Instance.changeOnline(true)
            PurpleLogger.current.d(RabbitMQBroker.TAG, "startConnection, connected!")
        }catch let error {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "startConnection, exception!")
        }
    }
    
    private func doUserChannelThings(_ userChanged: Bool, _ groupChanged: Bool) {
        if userChanged {
            fullDeclare()
        }else if groupChanged {
            someDeclare()
        }
    }
    
    private func someDeclare() {
        PurpleLogger.current.d(RabbitMQBroker.TAG, "someDeclare")
        if !checkConnection(){
            PurpleLogger.current.d(RabbitMQBroker.TAG, "someDeclare, connection not ready, return")
            return
        }
        guard let property = self.property else {
            return
        }
        guard let channel = self.channel else{
            return
        }
        guard let queueName = self.queueName else{
            return
        }
        let routingKey = "\(property.routingKeyPrefix)\(property.queuePrefix)\(uid ?? "")"
        var allExchanges:[Triple<String, String, String>] = []
        if let othersExchanges = fillExchanges(property) {
            allExchanges += othersExchanges
        }
        for exchange in allExchanges {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "someDeclare, exchange:\(exchange)")
            channel.exchangeDeclare(exchange.second, type: exchange.third)
            channel.queueBind(queueName, exchange: exchange.second, routingKey: routingKey)
            if exchange.third == RMQConstants.EXCHNAGE_FANOUT {
                channel.exchangeBind(property.groupExchangeParent, destination: exchange.second, routingKey: "msg.to.group_\(exchange.first)")
            }
        }
        
    }
    
    func fullDeclare() {
        PurpleLogger.current.d(RabbitMQBroker.TAG, "fullDeclare")
        if !checkConnection(){
            PurpleLogger.current.d(RabbitMQBroker.TAG, "fullDeclare, connection not ready, return")
            return
        }
        guard let property = self.property else {
            return
        }
        
        guard let channel = self.channel else {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "fullDeclare, channel is null, return")
            return
        }
        
        guard let uid = self.uid else {
            return
        }
        
        let queueName = "\(property.queuePrefix)\(uid)_CUUID_\(UUID().uuidString.lowercased())"
        
        self.queueName = queueName
        
        let routingKey = "\(property.routingKeyPrefix)\(property.queuePrefix)\(uid)"
        
        let mQueue = channel.queue(queueName, options: [.exclusive])
        
        channel.exchangeDeclare(property.groupExchangeParent, type: RMQConstants.EXCHNAGE_TOPIC, options: [.durable])
        channel.exchangeDeclare(property.groupSubRootExchange, type: RMQConstants.EXCHNAGE_FANOUT, options: [.durable])
        channel.exchangeBind(property.groupSubRootExchange, destination: property.groupExchangeParent, routingKey: "")
        
        
        var allExchanges: [Triple<String, String, String>] = []
        allExchanges.append(Triple(uid, "\(property.groupExchangePrefix)\(uid)", RMQConstants.EXCHNAGE_FANOUT))
        if let othersExchanges = fillExchanges(property) {
            allExchanges += othersExchanges
        }
        for exchange in allExchanges {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "fullDeclare, exchange:\(exchange)")
            channel.exchangeDeclare(exchange.second, type: exchange.third)
            channel.queueBind(queueName, exchange: exchange.second, routingKey: routingKey)
            if exchange.third == RMQConstants.EXCHNAGE_FANOUT {
                channel.exchangeBind(property.groupExchangeParent, destination: exchange.second, routingKey: "msg.to.group_\(exchange.first)")
            }
        }
        
        mQueue.subscribe([], handler: { message in
            PurpleLogger.current.d(RabbitMQBroker.TAG, "message received")
            guard message.body != nil else {
                PurpleLogger.current.d(RabbitMQBroker.TAG, "message received, content is null, return")
                return
            }
            
            if let imMessage = ImMessageGenerator.Instance.generateByReceived(message) {
                LocalMessager.post(RabbitMQBroker.MESSAGE_KEY_ON_IM_MESSAGE, imMessage)
            }
            
        })
    }
    
    
    func fillExchanges(_ property: RabbitMqBrokerProperty) -> [Triple<String, String, String>]? {
        let groupIds = RelationsUseCase.Instance.getGroupIds()
        if groupIds.isEmpty {
            return nil
        }
        
        var exchanges:[Triple<String, String, String>] = []
        
        for groupId in groupIds {
            let exchange = Triple(groupId, "\(property.groupExchangePrefix)\(groupId)", RMQConstants.EXCHNAGE_FANOUT)
            exchanges.append(exchange)
        }
        
        PurpleLogger.current.d(RabbitMQBroker.TAG, "fillExchanges, exchanges:\(exchanges)")
        
        return exchanges
    }
    
    
    func retry() {
    
    }
    
    func close() {
        
    }
    
    func checkConnection() -> Bool {
        guard let connection = self.connection else {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "checkConnection, broker connection is null, return false")
            return false
        }
        PurpleLogger.current.d(RabbitMQBroker.TAG, "checkConnection, broker connection isOpen:\(connection.isOpen())")
        //if !connection.isOpen() {
        //    PurpleLogger.current.d(RabbitMQBroker.TAG, "checkConnection, broker connection is not open, return false")
        //    return false
        //}
        
        guard let channel = self.channel else {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "checkConnection, broker connection channel is null, return false")
            return false
        }
        let isChannelOpen = channel.isOpen()
        PurpleLogger.current.d(RabbitMQBroker.TAG, "checkConnection, broker connection channel isOpen:\(isChannelOpen)")
        return isChannelOpen
    }
    
    private var brokerDeliveryMode = 2
    
    func sendMessage(_ imObj: any ImObj, _ imMessage: any ImMessage)-> Bool {
        if !checkConnection() {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "sendMessage, broker connection not ready! return")
            return false
        }
        PurpleLogger.current.d(RabbitMQBroker.TAG, "sendMessage")
        guard let channel = self.channel else {
            return false
        }
        guard let property = self.property else {
            return false
        }
        
        guard let data = ImMessageGenerator.Instance.makeMessageMetadataAsJsonStringData(message: imMessage) else {
            return false
        }
        
        var properties: [RMQValue] = []
        let timestamp = RMQBasicTimestamp(imMessage.timestamp)
        let type = RMQBasicType(ImMessageDesignType.getType(imMessage))
        let deliveryMode = RMQBasicDeliveryMode(CChar(brokerDeliveryMode))
        properties.append(timestamp)
        properties.append(type)
        properties.append(deliveryMode)
        var headers: [String: RMQValue & RMQFieldValue] = [String: RMQValue & RMQFieldValue]()
        headers[ImMessageConstant.HEADER_MESSAGE_MESSAGE_DELIVERY_TYPE] = RMQLongstr(AppSettings.Instance.imMessageDeliveryType)
        headers[ImMessageConstant.HEADER_MESSAGE_ID] = RMQLongstr(imMessage.id)
        headers[ImMessageConstant.HEADER_MESSAGE_UID] = RMQLongstr(imMessage.fromInfo.uid)
        headers[ImMessageConstant.HEADER_MESSAGE_NAME] = RMQLongstr(imMessage.fromInfo.name ?? "")
        let fromUserNameBase64 = (imMessage.fromInfo.name ?? "").data(using: .utf8)?.base64EncodedString() ?? ""
        headers[ImMessageConstant.HEADER_MESSAGE_NAME_BASE64] = RMQLongstr(fromUserNameBase64)
        headers[ImMessageConstant.HEADER_MESSAGE_AVATAR_URL] = RMQLongstr(imMessage.fromInfo.avatarUrl ?? "")
        headers[ImMessageConstant.HEADER_MESSAGE_ROLES] = RMQLongstr(imMessage.fromInfo.roles ?? "")
        headers[ImMessageConstant.HEADER_MESSAGE_MESSAGE_GROUP_TAG] = RMQLongstr(imMessage.messageGroupTag ?? "")
        headers[ImMessageConstant.HEADER_MESSAGE_TO_ID] = RMQLongstr(imMessage.toInfo.id)
        headers[ImMessageConstant.HEADER_MESSAGE_TO_NAME] = RMQLongstr(imMessage.toInfo.name ?? "")
        let toNameBase64 = (imMessage.toInfo.name ?? "").data(using: .utf8)?.base64EncodedString() ?? ""
        headers[ImMessageConstant.HEADER_MESSAGE_TO_NAME_BASE64] = RMQLongstr(toNameBase64)
        headers[ImMessageConstant.HEADER_MESSAGE_TO_TYPE] = RMQLongstr(imMessage.toInfo.toType)
        headers[ImMessageConstant.HEADER_MESSAGE_TO_ICON_URL] = RMQLongstr(imMessage.toInfo.iconUrl ?? "")
        headers[ImMessageConstant.HEADER_MESSAGE_TO_ROLES] = RMQLongstr(imMessage.toInfo.roles ?? "")
        let basicHeaders = RMQBasicHeaders(headers)
        properties.append(basicHeaders)
    
        let exchange: String = property.groupSubRootExchange
        
        let routingKey: String = "msg.to.group_\(imObj.id)"
        
        do{
            try channel.basicPublish(data, routingKey: routingKey, exchange: exchange, properties: properties)
            PurpleLogger.current.d(RabbitMQBroker.TAG, "sendMessage, exchange:\(exchange), routingKey:\(routingKey)")
        }catch let error {
            PurpleLogger.current.d(RabbitMQBroker.TAG, "sendMessage, exception! error\(error)")
            return false
        }
        
        
        
        return true
    }
    
    func updateImGroupBindIfNeeded() {
        PurpleLogger.current.d(RabbitMQBroker.TAG, "updateImGroupBindIfNeeded")
        guard let property = property else {
            return
        }
        updateUserChannel(property)
    }
}

File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/im/RabbitMQBroker.kt
package xcj.app.appsets.im

import com.rabbitmq.client.AMQP
import com.rabbitmq.client.BlockedListener
import com.rabbitmq.client.BuiltinExchangeType
import com.rabbitmq.client.CancelCallback
import com.rabbitmq.client.Channel
import com.rabbitmq.client.ConnectionFactory
import com.rabbitmq.client.DeliverCallback
import com.rabbitmq.client.Delivery
import com.rabbitmq.client.Recoverable
import com.rabbitmq.client.RecoveryListener
import com.rabbitmq.client.impl.recovery.AutorecoveringConnection
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import xcj.app.appsets.account.LocalAccountManager
import xcj.app.appsets.im.message.IMMessage
import xcj.app.appsets.im.message.MessageSendInfo
import xcj.app.appsets.settings.AppSetsModuleSettings
import xcj.app.appsets.usecase.ConversationUseCase
import xcj.app.appsets.usecase.RelationsUseCase
import xcj.app.starter.android.util.LocalMessenger
import xcj.app.starter.android.util.PurpleLogger
import xcj.app.starter.test.LocalApplication
import xcj.app.starter.test.LocalPurpleCoroutineScope
import java.util.UUID
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

class RabbitMQBroker : MessageBroker<RabbitMQBrokerConfig>,
    DeliverCallback, CancelCallback,
    RecoveryListener, BlockedListener {

    companion object {
        private const val TAG = "RabbitMQBroker"
    }

    private lateinit var property: RabbitMQBrokerProperty

    private var connection: AutorecoveringConnection? = null

    private var channel: Channel? = null

    private var queueName: String? = null

    private var uid: String? = null

    private var brokerDeliveryMode = 2//1->non-persistent, 2->persistent

    fun isConnected(): Boolean {
        return checkConnection()
    }

    override suspend fun retry() {

    }

    /**
     * start point
     */
    override suspend fun bootstrap(config: RabbitMQBrokerConfig) {
        return updateUserChannel(config.getRabbitProperty())
    }

    private suspend fun updateUserChannel(property: RabbitMQBrokerProperty) {
        PurpleLogger.current.d(TAG, "updateUserChannel")
        withContext(Dispatchers.IO) {
            this@RabbitMQBroker.property = property
            val userGroupsChanged = true
            val loggedUserUid = LocalAccountManager.userInfo.uid
            val userChanged = uid != loggedUserUid
            uid = loggedUserUid
            basicConnect()
            handleUserChannel(userChanged, userGroupsChanged)
        }
    }

    suspend fun updateImGroupBindIfNeeded() {
        PurpleLogger.current.d(TAG, "updateImGroupBindIfNeeded")
        if (!::property.isInitialized) {
            return
        }
        updateUserChannel(property)
    }

    override suspend fun close() {
        withContext(Dispatchers.IO) {
            deleteUserQueue()
            closeConnection()
        }
    }

    private fun deleteUserQueue() {
        if (!checkConnection()) {
            return
        }
        if (queueName.isNullOrEmpty()) {
            return
        }
        runCatching {
            channel?.queueDelete(queueName)
        }.onSuccess {
            PurpleLogger.current.d(TAG, "deleteUserQueue, success")
        }.onFailure {
            PurpleLogger.current.d(TAG, "deleteUserQueue, failed")
        }
    }

    private fun closeConnection() {
        runCatching {
            connection?.removeBlockedListener(this)
            connection?.removeRecoveryListener(this)
            channel?.close()
            connection?.close()
            channel = null
            connection = null
            queueName = null
            uid = null
        }.onSuccess {
            BrokerTest.imOnLineState.value = IMOnlineState.Offline
            PurpleLogger.current.d(TAG, "closeConnection, success")
        }.onFailure {
            PurpleLogger.current.d(TAG, "closeConnection, failed")
        }
    }

    override fun handleBlocked(reason: String?) {
        PurpleLogger.current.d(TAG, "handleBlocked, reason:$reason")
    }

    override fun handleUnblocked() {
        PurpleLogger.current.d(TAG, "handleUnblocked")
    }

    override fun handleRecovery(recoverable: Recoverable?) {
        PurpleLogger.current.d(TAG, "handleRecovery")
        BrokerTest.imOnLineState.value = IMOnlineState.Online()
    }

    override fun handleRecoveryStarted(recoverable: Recoverable?) {
        PurpleLogger.current.d(TAG, "handleRecoveryStarted")
        BrokerTest.imOnLineState.value = IMOnlineState.Offline
    }

    private fun basicConnect() {
        if (checkConnection()) {
            PurpleLogger.current.d(
                TAG,
                "basicConnect! connection is connected, return"
            )
            return
        }
        PurpleLogger.current.d(
            TAG,
            "basicConnect, start, rabbit host:port: ${property.host}:${property.port}"
        )

        val factory = ConnectionFactory().apply {
            host = property.host
            port = property.port
            username = property.username
            password = property.password
            virtualHost = property.virtualHost
            useNio()
            setNetworkRecoveryInterval(500)//millis seconds, 0.5s
            //requestedHeartbeat = 10//seconds
        }
        runCatching {
            val autoRecoveringConnection = factory.newConnection() as AutorecoveringConnection
            autoRecoveringConnection.addRecoveryListener(this)
            autoRecoveringConnection.addBlockedListener(this)
            connection = autoRecoveringConnection
            channel = autoRecoveringConnection.createChannel()
        }.onSuccess {
            BrokerTest.imOnLineState.value = IMOnlineState.Online()
            PurpleLogger.current.d(
                TAG,
                "basicConnect, final connection connected!"
            )
        }.onFailure {
            PurpleLogger.current.d(
                TAG,
                "basicConnect, exception! ${it.message}"
            )
        }
    }

    private fun handleUserChannel(userChanged: Boolean, groupsChanged: Boolean) {
        if (userChanged) {
            runCatching {
                fullDeclare()
            }.onFailure {
                PurpleLogger.current.d(
                    TAG,
                    "handleUserChannel, fullDeclare exception! ${it.message}"
                )
            }
        } else if (groupsChanged) {
            runCatching {
                someDeclare()
            }.onFailure {
                PurpleLogger.current.d(
                    TAG,
                    "handleUserChannel, someDeclare exception! ${it.message}"
                )
            }
        }
    }

    private fun checkConnection(): Boolean {
        val autoRecoveringConnection = connection
        if (autoRecoveringConnection == null) {
            PurpleLogger.current.d(TAG, "checkConnection, connection is null, return")
            return false
        }
        if (!autoRecoveringConnection.isOpen) {
            PurpleLogger.current.d(TAG, "checkConnection, connection is not open, return")
            return false
        }
        val channel = channel
        if (channel == null) {
            PurpleLogger.current.d(TAG, "checkConnection, channel is null, return")
            return false
        }
        if (!channel.isOpen) {
            PurpleLogger.current.d(TAG, "checkConnection, channel is not open, return")
            return false
        }
        return true
    }

    private fun someDeclare() {
        if (!checkConnection()) {
            PurpleLogger.current.d(TAG, "someDeclare, connection not ready， return")
            return
        }
        PurpleLogger.current.d(TAG, "someDeclare, start")
        val channel = channel ?: return
        val queueName = this.queueName ?: return
        val routingKey = "${property.routingKeyPrefix}${property.queuePrefix}${uid}"
        val allExchanges: MutableList<Triple<String, String, BuiltinExchangeType>> = mutableListOf()

        RelationsUseCase.getInstance().getGroupIds().forEach {
            val exchange =
                Triple(it, "${property.groupExchangePrefix}$it", BuiltinExchangeType.FANOUT)
            allExchanges.add(exchange)
        }

        allExchanges.forEach { exchange ->
            PurpleLogger.current.d(
                TAG,
                "someDeclare, exchange:${exchange.second}, type:${exchange.third}"
            )
            channel.exchangeDeclare(exchange.second, exchange.third, false, false, null)
            channel.queueBind(queueName, exchange.second, routingKey)
            if (exchange.third == BuiltinExchangeType.FANOUT) {
                channel.exchangeBind(
                    exchange.second,
                    property.groupExchangeParent,
                    "msg.to.group_${exchange.first}"
                )
            }
        }
        PurpleLogger.current.d(TAG, "someDeclare, final")
    }

    private fun fullDeclare() {
        if (!checkConnection()) {
            PurpleLogger.current.d(TAG, "fullDeclare, connection not ready, return")
            return
        }
        PurpleLogger.current.d(TAG, "fullDeclare, start")
        val channel = channel ?: return
        val uid = uid ?: return
        val queueName = "${property.queuePrefix}${uid}_CUUID_${UUID.randomUUID()}"
        this.queueName = queueName
        val routingKey = "${property.routingKeyPrefix}${property.queuePrefix}${uid}"

        channel.queueDeclare(queueName, false, true, false, null)

        channel.exchangeDeclare(property.groupExchangeParent, BuiltinExchangeType.TOPIC, true)

        channel.exchangeDeclare(property.groupSubRootExchange, BuiltinExchangeType.FANOUT, true)


        channel.exchangeBind(property.groupExchangeParent, property.groupSubRootExchange, "")

        val allExchanges: MutableList<Triple<String, String, BuiltinExchangeType>> = mutableListOf()
        allExchanges.add(
            Triple(
                uid,
                "${property.groupExchangePrefix}$uid",
                BuiltinExchangeType.FANOUT
            )
        )
        RelationsUseCase.getInstance().getGroupIds().forEach {
            val exchange =
                Triple(it, "${property.groupExchangePrefix}$it", BuiltinExchangeType.FANOUT)
            allExchanges.add(exchange)
        }

        allExchanges.forEach { exchange ->
            PurpleLogger.current.d(
                TAG,
                "fullDeclare, exchange:${exchange.second}, type:${exchange.third}"
            )
            channel.exchangeDeclare(exchange.second, exchange.third, false, false, null)
            channel.queueBind(queueName, exchange.second, routingKey)
            if (exchange.third == BuiltinExchangeType.FANOUT) {
                channel.exchangeBind(
                    exchange.second,
                    property.groupExchangeParent,
                    "msg.to.group_${exchange.first}"
                )
            }
        }

        channel.basicConsume(queueName, true, this, this)
        PurpleLogger.current.d(TAG, "fullDeclare, final")
    }

    override fun handle(consumerTag: String?) {
        PurpleLogger.current.d(TAG, "message consume cancel, consume tag is:$consumerTag!")
    }

    override fun handle(consumerTag: String?, message: Delivery?) {
        PurpleLogger.current.d(
            TAG,
            "message delivery from rabbitmq!, consume tag is:$consumerTag, message:$message"
        )
        if (message == null) {
            return
        }
        LocalPurpleCoroutineScope.current.launch {
            IMMessageGenerator.generateByReceived(message)?.let { imMessage ->
                ConversationUseCase.getInstance()
                    .onMessage(LocalApplication.current, imMessage, false)
                LocalMessenger.post(MessageBrokerConstants.MESSAGE_KEY_ON_IM_MESSAGE, imMessage)
            }
        }
    }

    @OptIn(ExperimentalEncodingApi::class)
    override suspend fun sendMessage(imObj: IMObj, imMessage: IMMessage) {
        if (!checkConnection()) {
            PurpleLogger.current.d(
                TAG,
                "sendMessage, broker connection not ready! return"
            )
            imMessage.updateSending(
                MessageSendInfo(
                    isSent = false,
                    failureReason = "connection not ready!"
                )
            )
            return
        }
        val channel = channel
        if (channel == null) {
            imMessage.updateSending(
                MessageSendInfo(
                    isSent = false,
                    failureReason = "channel not ready!"
                )
            )
            return
        }

        withContext(Dispatchers.IO) {
            PurpleLogger.current.d(TAG, "sendMessage")
            val properties = AMQP.BasicProperties.Builder().apply {
                timestamp(imMessage.timestamp)
                type(imMessage.messageType)
                deliveryMode(brokerDeliveryMode)

                val headers = mutableMapOf<String, Any?>()
                headers[IMMessage.HEADER_MESSAGE_MESSAGE_DELIVERY_TYPE] =
                    AppSetsModuleSettings.get().imMessageDeliveryType
                headers[IMMessage.HEADER_MESSAGE_ID] = imMessage.id
                headers[IMMessage.HEADER_MESSAGE_UID] = imMessage.fromInfo.uid
                headers[IMMessage.HEADER_MESSAGE_NAME] = imMessage.fromInfo.bioName
                headers[IMMessage.HEADER_MESSAGE_NAME_BASE64] =
                    imMessage.fromInfo.bioName?.let {
                        Base64.encode(it.toByteArray())
                    }
                headers[IMMessage.HEADER_MESSAGE_AVATAR_URL] =
                    imMessage.fromInfo.avatarUrl
                headers[IMMessage.HEADER_MESSAGE_ROLES] = imMessage.fromInfo.roles

                if (!imMessage.messageGroupTag.isNullOrEmpty()) {
                    headers[IMMessage.HEADER_MESSAGE_MESSAGE_GROUP_TAG] =
                        imMessage.messageGroupTag
                }
                if (imMessage.toInfo.bioId.isNotEmpty()) {
                    headers[IMMessage.HEADER_MESSAGE_TO_ID] = imMessage.toInfo.bioId
                }
                if (!imMessage.toInfo.bioName.isNullOrEmpty()) {
                    headers[IMMessage.HEADER_MESSAGE_TO_NAME] = imMessage.toInfo.bioName
                    headers[IMMessage.HEADER_MESSAGE_TO_NAME_BASE64] =
                        imMessage.toInfo.bioName?.let {
                            Base64.encode(it.toByteArray())
                        }
                }
                if (imMessage.toInfo.toType.isNotEmpty()) {
                    headers[IMMessage.HEADER_MESSAGE_TO_TYPE] = imMessage.toInfo.toType
                }
                if (!imMessage.toInfo.iconUrl.isNullOrEmpty()) {
                    headers[IMMessage.HEADER_MESSAGE_TO_ICON_URL] =
                        imMessage.toInfo.iconUrl
                }
                if (!imMessage.toInfo.roles.isNullOrEmpty()) {
                    headers[IMMessage.HEADER_MESSAGE_TO_ROLES] = imMessage.toInfo.roles
                }

                headers(headers)
            }.build()
            val exchange = property.groupSubRootExchange
            val routingKey = "msg.to.group_${imObj.id}"
            val contentBytes =
                IMMessageGenerator.makeMessageMetadataAsJsonString(imMessage).toByteArray()
            PurpleLogger.current.d(
                TAG,
                "sendMessage, exchange:$exchange, routingKey:$routingKey"
            )
            runCatching {
                channel.basicPublish(exchange, routingKey, properties, contentBytes)
                imMessage.updateSending(MessageSendInfo(isSent = true))
            }.onFailure {
                PurpleLogger.current.e(TAG, "sendMessage, failed, ${it.message}")
                imMessage.updateSending(
                    MessageSendInfo(
                        isSent = false,
                        failureReason = it.localizedMessage
                    )
                )
            }
        }
    }
}
